From b60b322cf9f0ffa624018f1bb9783edf0dc908c3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?J=C3=B6rn=20Heusipp?= <manx@openmpt.org>
Date: Sat, 3 Feb 2018 16:14:38 +0000
Subject: [PATCH] [Sec] STP: Possible out-of-bounds memory read with malformed
 STP files (caught with afl-fuzz). Patch-by: sagamusix (originally committed
 as part of r9568)

git-svn-id: https://source.openmpt.org/svn/openmpt/branches/OpenMPT-1.27@9576 56274372-70c3-4bfc-bfc3-4c3a0b034d27
---
 soundlib/Load_stp.cpp | 13 +++++++++----
 1 file changed, 9 insertions(+), 4 deletions(-)

diff --git a/soundlib/Load_stp.cpp b/soundlib/Load_stp.cpp
index 68219a29f9..4389a4f357 100644
--- a/soundlib/Load_stp.cpp
+++ b/soundlib/Load_stp.cpp
@@ -108,7 +108,12 @@ static TEMPO ConvertTempo(uint16 ciaSpeed)
 
 static void ConvertLoopSlice(ModSample &src, ModSample &dest, SmpLength start, SmpLength len, bool loop)
 {
-	if(!src.HasSampleData()) return;
+	if(!src.HasSampleData()
+		|| start >= src.nLength
+		|| src.nLength - start < len)
+	{
+		return;
+	}
 
 	dest.FreeSample();
 	dest = src;
@@ -156,9 +161,9 @@ static void ConvertLoopSequence(ModSample &smp, STPLoopList &loopList)
 
 		// If adding this loop would cause the sample length to exceed maximum,
 		// then limit and bail out
-		if((newSmp.nLength + info.loopLength > MAX_SAMPLE_LENGTH) ||
-		   (info.loopLength > MAX_SAMPLE_LENGTH) ||
-		   (info.loopStart + info.loopLength > smp.nLength))
+		if(info.loopStart >= smp.nLength
+			|| smp.nLength - info.loopStart < info.loopLength
+			|| newSmp.nLength > MAX_SAMPLE_LENGTH - info.loopLength)
 		{
 			numLoops = i;
 			break;
